include/xtensor/xaxis_slice_iterator.hpp

```diff
@@ -0,0 +1,226 @@
+        mutable storing_type p_expression;
+        size_type m_index;
+        size_type m_offset;
+        size_type m_axis_stride;
+        size_type m_lower_shape;
+        size_type m_upper_shape;
+        size_type m_iter_size;
+        bool m_is_target_axis;
+        value_type m_sv;
+
+        template <class T, class CTA>
+        std::enable_if_t<std::is_pointer<T>::value, T>
+            get_storage_init(CTA&& e) const;
+
+        template <class T, class CTA>
+        std::enable_if_t<!std::is_pointer<T>::value, T>
+            get_storage_init(CTA&& e) const;
+    };
+
+    template <class CT>
+    bool operator==(const xaxis_slice_iterator<CT>& lhs, const xaxis_slice_iterator<CT>& rhs);
+
+    template <class CT>
+    bool operator!=(const xaxis_slice_iterator<CT>& lhs, const xaxis_slice_iterator<CT>& rhs);
+
+    template <class E>
+    auto xaxis_slice_begin(E&& e);
+
+    template <class E>
+    auto xaxis_slice_begin(E&& e, typename std::decay_t<E>::size_type axis);
+
+    template <class E>
+    auto xaxis_slice_end(E&& e);
+
+    template <class E>
+    auto xaxis_slice_end(E&& e, typename std::decay_t<E>::size_type axis);
+
+    /***************************************
+     * xaxis_slice_iterator implementation *
+     ***************************************/
+
+    template <class CT>
+    template <class T, class CTA>
+    inline std::enable_if_t<std::is_pointer<T>::value, T>
+        xaxis_slice_iterator<CT>::get_storage_init(CTA&& e) const
+    {
+        return &e;
+    }
+
+    template <class CT>
+    template <class T, class CTA>
+    inline std::enable_if_t<!std::is_pointer<T>::value, T>
+        xaxis_slice_iterator<CT>::get_storage_init(CTA&& e) const
+    {
+        return e;
+    }
+
+    template <class CT>
+    template <class CTA>
+    inline xaxis_slice_iterator<CT>::xaxis_slice_iterator(CTA&& e, size_type axis)
+        : xaxis_slice_iterator(std::forward<CTA>(e), axis, 0, e.data_offset())
+    {
+    }
+
+    template <class CT>
+    template <class CTA>
+    inline xaxis_slice_iterator<CT>::xaxis_slice_iterator(CTA&& e, size_type axis, size_type index, size_type offset) :
+        p_expression(get_storage_init<storing_type>(std::forward<CTA>(e))), m_index(index),
+        m_offset(offset), m_axis_stride(e.strides()[axis] * (e.shape()[axis] - 1)),
+        m_lower_shape(0), m_upper_shape(0), m_iter_size(0), m_is_target_axis(false),
+        m_sv(strided_view(std::forward<CT>(e), std::forward<shape_type>({ e.shape()[axis] }),
+            std::forward<strides_type>({ e.strides()[axis] }), offset, e.layout()))
+    {
+        if (e.layout() == layout_type::row_major)
+        {
+            m_is_target_axis = axis == e.dimension() - 1;
+            m_lower_shape = std::accumulate(e.shape().begin() + axis + 1, e.shape().end(), static_cast<size_t>(1), std::multiplies<>());
+            m_iter_size = std::accumulate(e.shape().begin() + 1, e.shape().end(), static_cast<size_t>(1), std::multiplies<>());
```

Comment:
nitpicking: the `static_cast` can be omitted here to alleviate the code: `size_t(1)`

Reply:
Changed to use size_t(1).